plugin-cgroupv2/src/k8s/token.rs

```diff
@@ -0,0 +1,140 @@
+
+pub struct Token {
+    expiration_time: RwLock<Option<u64>>,
+    token_retrieval: TokenRetrieval,
+    value: RwLock<Option<String>>,
+}
+
+impl Token {
+    pub fn new(token_retrieval: TokenRetrieval) -> Self {
+        Self {
+            expiration_time: RwLock::new(None),
+            token_retrieval,
+            value: RwLock::new(None),
+        }
+    }
+
+    /// Check if the token' lifetime has reached it's expiration or not.
+    async fn is_valid(&self) -> bool {
+        let now_secs = match SystemTime::now().duration_since(UNIX_EPOCH) {
+            Ok(ts) => ts.as_secs(),
+            Err(_) => return false,
+        };
+
+        self.expiration_time.read().await.is_some()
+            && self.value.read().await.is_some()
+            && now_secs < self.expiration_time.read().await.unwrap_or_default() // to prevent any weird panic (won't happen)
+    }
+
+    /// Returns the current value of the token if it is still alive or
+    /// it's refreshed value if not.
+    pub async fn get_value(self: Arc<Self>) -> anyhow::Result<String> {
+        if self.is_valid().await {
+            return match self.value.read().await.clone() {
+                Some(value) => Ok(value),
+                None => Err(anyhow!("could no read the token's value")),
+            };
+        }
+
+        self.refresh().await
+    }
+
+    /// Retrieves the k8s API token using either a kubectl command
+    /// or by reading  the service account token's file.
+    async fn refresh(self: Arc<Self>) -> anyhow::Result<String> {
+        let token = match self.token_retrieval {
+            TokenRetrieval::Kubectl => {
+                let output = Command::new("kubectl")
+                    .args(["create", "token", "alumet-reader", "-n", "alumet"])
+                    .output()?;
+
+                let token = String::from_utf8_lossy(&output.stdout);
+                token.trim().to_string()
+            }
+            TokenRetrieval::File => {
+                let mut file = File::open("/var/run/secrets/kubernetes.io/serviceaccount/token")?;
+                let mut token = String::new();
+                file.read_to_string(&mut token)?;
+                token
+            }
+        };
+
+        let mut components = token.split('.');
+        let _ = components.next().ok_or(anyhow!("Could not parse the token"))?;
+        let payload: serde_json::Value = serde_json::from_slice(
+            &BASE64_STANDARD_NO_PAD.decode(components.next().ok_or(anyhow!("Could not parse the token"))?.trim())?,
+        )?;
+        let _ = components.next().ok_or(anyhow!("Could not parse the token"))?;
+
+        if components.next().is_some() {
+            return Err(anyhow!("Token has too many components"));
+        }
+
+        // Extract 'exp' from the payload
+        let exp = match payload.get("exp") {
+            Some(exp) => exp.as_u64().unwrap_or_default(),
+            None => return Err(anyhow!("Could not extract the 'exp' field from the token")),
```

Comment:
The JWT `exp` field is optional. What happens when a token is generated without an expiration date?
IIUC the expiration date should be set to None in that case.

Reply:
It is optional in the JWT structure definition, but Kubernetes is using it in any case
But yes I can change this, you're right